Integrate setTimeout and setInterval with HTML5 event loop

rniwa opened the pull request:
#### 43d0b6dbeebf8116a9ff8af86280528bd8884676
<pre>
Integrate setTimeout and setInterval with HTML5 event loop
<a href="https://bugs.webkit.org/show_bug.cgi?id=203137">https://bugs.webkit.org/show_bug.cgi?id=203137</a>

Reviewed by Chris Dumez.

This PR integrates DOM timers with HTML5 event loop. Specifically, DOMTimer now uses
EventLoopTaskGroup::scheduleTask and EventLoopTaskGroup::scheduleRepeatingTask instead of
inheriting from SuspendableTimer. This PR also deletes SuspendableTimer class now that
we&apos;ve migrated all uses of it to EventLoop&apos;s equivalents.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/EventLoop.cpp:
(WebCore::EventLoop::scheduleTask): Fixed a bug that the newly created timer won&apos;t be
suspended when the task group had been suspended.
(WebCore::EventLoop::scheduleRepeatingTask): Ditto.
(WebCore::EventLoopTaskGroup::setTimerAlignment):
(WebCore::EventLoopTaskGroup::setTimerHasReachedMaxNestingLevel):
(WebCore::EventLoopTaskGroup::adjustTimerNextTimeout):
(WebCore::EventLoopTaskGroup::adjustTimerRepeatInterval):
(WebCore::EventLoopTaskGroup::didChangeTimerAlignmentInterval):
* Source/WebCore/dom/EventLoop.h:
* Source/WebCore/dom/ScriptExecutionContext.cpp:
(WebCore::ScriptExecutionContext::didChangeTimerAlignmentInterval):
* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/dom/TaskSource.h:
* Source/WebCore/page/DOMTimer.cpp:
(WebCore::DOMTimer::DOMTimer):
(WebCore::DOMTimer::removeById):
(WebCore::DOMTimer::fired):
(WebCore::DOMTimer::stop):
(WebCore::DOMTimer::updateTimerIntervalIfNecessary):
(WebCore::DOMTimer::didStop): Deleted.
* Source/WebCore/page/DOMTimer.h:
* Source/WebCore/page/SuspendableTimer.cpp: Removed.
* Source/WebCore/page/SuspendableTimer.h: Removed.
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::isTimerThrottled):

Canonical link: <a href="https://commits.webkit.org/266828@main">https://commits.webkit.org/266828@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1839804d6ddbe47aee0ae01e9eccfaec61d0afc0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14824 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15129 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15481 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16571 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13953 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14961 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17650 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15229 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/16598 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15004 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15497 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12590 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17305 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12776 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13375 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/20354 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13855 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13543 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16783 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14089 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11908 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13380 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17711 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1781 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13933 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/dom/EventLoop.h:
```diff
@@ -208,6 +210,12 @@ class EventLoopTaskGroup : public CanMakeWeakPtr<EventLoopTaskGroup> {
+    void setTimerAlignment(EventLoopTimerHandle, TimerAlignment&);
+    void didChangeTimerAlignmentInterval(EventLoopTimerHandle);
+    void setTimerHasReachedMaxNestingLevel(EventLoopTimerHandle, bool);
+    void augmentTimerNextTimeout(EventLoopTimerHandle, Seconds delta);
```
Not sure that "augment" is the best word: I have to stop and think about whether it means "replace" or "extend". Maybe "extend" is the right term to use?

Also, "timeout" => "fire time".

rniwa replied:
Yeah, adjustTimerFireTime?

cdumez reviewed (approved):
LGTM

rniwa committed:
Integrate setTimeout and setInterval with HTML5 event loop
https://bugs.webkit.org/show_bug.cgi?id=203137

Reviewed by Chris Dumez.

This PR integrates DOM timers with HTML5 event loop. Specifically, DOMTimer now uses
EventLoopTaskGroup::scheduleTask and EventLoopTaskGroup::scheduleRepeatingTask instead of
inheriting from SuspendableTimer. This PR also deletes SuspendableTimer class now that
we've migrated all uses of it to EventLoop's equivalents.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/EventLoop.cpp:
(WebCore::EventLoop::scheduleTask): Fixed a bug that the newly created timer won't be
suspended when the task group had been suspended.
(WebCore::EventLoop::scheduleRepeatingTask): Ditto.
(WebCore::EventLoopTaskGroup::setTimerAlignment):
(WebCore::EventLoopTaskGroup::setTimerHasReachedMaxNestingLevel):
(WebCore::EventLoopTaskGroup::adjustTimerNextTimeout):
(WebCore::EventLoopTaskGroup::adjustTimerRepeatInterval):
(WebCore::EventLoopTaskGroup::didChangeTimerAlignmentInterval):
* Source/WebCore/dom/EventLoop.h:
* Source/WebCore/dom/ScriptExecutionContext.cpp:
(WebCore::ScriptExecutionContext::didChangeTimerAlignmentInterval):
* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/dom/TaskSource.h:
* Source/WebCore/page/DOMTimer.cpp:
(WebCore::DOMTimer::DOMTimer):
(WebCore::DOMTimer::removeById):
(WebCore::DOMTimer::fired):
(WebCore::DOMTimer::stop):
(WebCore::DOMTimer::updateTimerIntervalIfNecessary):
(WebCore::DOMTimer::didStop): Deleted.
* Source/WebCore/page/DOMTimer.h:
* Source/WebCore/page/SuspendableTimer.cpp: Removed.
* Source/WebCore/page/SuspendableTimer.h: Removed.
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::isTimerThrottled):

Canonical link: https://commits.webkit.org/266828@main
